### PR TITLE
Update use_deterministic_algorithms docs

### LINKED ISSUE #55086: `use_deterministic_algorithms` documentation needs an update
There are a few issues with the documentation for `use_deterministic_algorithms`:

* Variable name `d` should be changed to something more descriptive
*  ":class:RuntimeError" is not displayed correctly
* Links to `index_put` and `__getitem__` don't work
* "self" should not be mentioned in user facing docs
* Language for backward functions should be fixed (#51487)
* Need to add a note saying that operations that cause multiple writes to the same location due to overlapping tensors or duplicate indices are not detected by nondeterministic alerts (#50750). Add a few examples of this

cc @mruberry @kurtamohler @brianjo

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #55086

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -367,46 +367,46 @@ def use_deterministic_algorithms(d):
         * :class:`torch.nn.ConvTranspose2d` when called on CUDA tensor
         * :class:`torch.nn.ConvTranspose3d` when called on CUDA tensor
         * :func:`torch.bmm` when called on sparse-dense CUDA tensors
-        * :func:`torch.__getitem__` backward when `self` is a CPU tensor and
-          ``indices`` is a list of tensors
-        * :func:`torch.index_put` with ``accumulate=True`` when called on a CPU
+        * :func:`torch.Tensor.__getitem__` automatic differentiation with CPU tensors
```
Comment (kurtamohler) [inline torch/__init__.py:370]: Note: This link still doesn't work, but I think that's fine because `__getitem__`/`index` does not seem to have an entry in the docs

File: torch/__init__.py [hunk 2]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
+        # Determinism is not guaranteed for an in-place operation on a tensor
+        # with internal memory overlap
+        >>> a = torch.tensor(1.).as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]),), torch.tensor([0., 1., 2.]))
+        tensor([2., 2., 2.])
+        >>> a = torch.tensor(1., device='cuda').as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]).cuda(),), torch.tensor([0., 1., 2.]).cuda())
+        tensor([0., 0., 0.], device='cuda:0')
+
+        # Determinism is not guaranteed for duplicate indices
```
Comment (kurtamohler) [inline torch/__init__.py:452]: @mruberry, since our video call, the plan has changed slightly. If I understand correctly, we should actually guarantee determinism for this case, by alerting without checking for duplicate indices. I can add an alert to `index_put_` if so. `embedding_bag` would also need one. We might have quite a few indexed functions that need an alert.

File: torch/__init__.py [hunk 3]
```diff
@@ -367,46 +367,46 @@ def use_deterministic_algorithms(d):
...
-        * :class:`torch.nn.AdaptiveAvgPool2d` when called on a CUDA tensor that requires grad
-        * :class:`torch.nn.AdaptiveAvgPool3d` when called on a CUDA tensor that requires grad
-        * :class:`torch.nn.MaxPool3d` when called on a CUDA tensor that requires grad
-        * :class:`torch.nn.AdaptiveMaxPool2d` when called on a CUDA tensor that requires grad
-        * :class:`torch.nn.FractionalMaxPool2d` when called on a CUDA tensor that requires grad
-        * :class:`torch.nn.FractionalMaxPool3d` when called on a CUDA tensor that requires grad
-        * :func:`torch.nn.functional.interpolate` when called on a CUDA tensor that requires grad
+    :class:`RuntimeError` when `mode=True`:
+
+        * :class:`torch.nn.AvgPool3d` automatic differentiation with CUDA tensors
```
Comment (kurtamohler) [inline torch/__init__.py:380]: @mruberry, what do you think of the way I phrased this?

Comment (mruberry) [reply torch/__init__.py:380]: I like it. I think it needs a tweak for its language to be parallel with the "with ..." language above. What about?

"... when attempting to differentiate a CUDA tensor"?

File: torch/__init__.py [hunk 4]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
+        nondeterministic behavior that is caused by an operation performing
+        multiple writes of different data to a single memory location. The
+        order of writes to a single location is not guaranteed, which can cause
+        nondeterminism. This could occur when an output tensor contains internal
+        memory overlaps or if an indexed operation is given duplicate indices.
```
Comment (mruberry) [inline torch/__init__.py:436]: I thought we were going to conservatively warn to prevent nondeterminism from duplicate indices?

File: torch/__init__.py [hunk 5]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
     Note that deterministic operations tend to have worse performance than
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
```
Comment (mruberry) [inline torch/__init__.py:431]: If the indexing operations part of this note is removed, should this just warn about internal memory overlaps? Like:

"Inplace operations on tensors that index the same element multiple times may not be detected or prevented by this flag."

Follow-up question, do we consider different orders of floating point or complex arithmetic operations to be nondeterministic?

File: torch/__init__.py [hunk 6]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
+        Below are several examples.
+
+    .. code-block:: python
+
+        >>> torch.use_deterministic_algorithms(True)
+
+        # Determinism is not guaranteed for an in-place operation on a tensor
+        # with internal memory overlap
+        >>> a = torch.tensor(1.).as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]),), torch.tensor([0., 1., 2.]))
```
Comment (mruberry) [inline torch/__init__.py:446]: An inplace add instead of index_put_ might demonstrate this point better since add is a simpler operation

File: torch/__init__.py (hunk 2 above)
Comment (mruberry) [reply torch/__init__.py:452]: Agreed. Filing issues for those functions would be great; if we can fix them before the PyTorch 1.9 release even better!

File: torch/__init__.py [hunk 7]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
+        >>> a = torch.zeros(3)
+        >>> a.index_put_((torch.tensor([1, 1, 1]),), torch.tensor([1., 2., 3.]))
+        tensor([0., 3., 0.])
+        >>> a.cuda().index_put_((torch.tensor([1, 1, 1]).cuda(),), torch.tensor([1., 2., 3.]).cuda())
+        tensor([0., 1., 0.], device='cuda:0')
+
     Args:
-        d (:class:`bool`): If True, force operations to be deterministic.
-                           If False, allow non-deterministic operations.
+        mode (:class:`bool`): If True, force operations to be deterministic.
```
Comment (mruberry) [inline torch/__init__.py:460]: 'force' -> 'forces'
'allow' -> 'allows'

Conceptually it's the mode taking action, so the mode "forces" the operations. 

What about elaborating slightly in the first case?

"If True, makes potentially nondeterministic operations switch to a deterministic algorithm or throw a runtime error.

File: torch/__init__.py [hunk 8]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
+        order of writes to a single location is not guaranteed, which can cause
+        nondeterminism. This could occur when an output tensor contains internal
+        memory overlaps or if an indexed operation is given duplicate indices.
+        Below are several examples.
+
+    .. code-block:: python
+
+        >>> torch.use_deterministic_algorithms(True)
+
+        # Determinism is not guaranteed for an in-place operation on a tensor
```
Comment (mruberry) [inline torch/__init__.py:443]: We might as well add the examples in the revision of this PR (if they're a pain we can always separate them to another PR). Example ideas from our notes:

- forward
- backward
- overlapped tensor
- writing to multiple locations (which will throw a nondeterminism error)

File: torch/__init__.py [hunk 9]
```diff
@@ -345,20 +345,20 @@ def set_default_dtype(d):
...
+def use_deterministic_algorithms(mode):
     r""" Sets whether PyTorch operations must use "deterministic"
     algorithms. That is, algorithms which, given the same input, and when
     run on the same software and hardware, always produce the same output.
     When True, operations will use deterministic algorithms when available,
     and if only nondeterministic algorithms are available they will throw a
-    :class:RuntimeError when called.
+    :class:`RuntimeError` when called.
 
     .. warning::
```
Comment (mruberry) [inline torch/__init__.py:356]: I suppose we should wait to rip this label off until the feature review committee agrees ;)

File: torch/__init__.py [hunk 10]
```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
...
+        tensor([0., 0., 0.], device='cuda:0')
+
+        # Determinism is not guaranteed for duplicate indices
+        >>> a = torch.zeros(3)
+        >>> a.index_put_((torch.tensor([1, 1, 1]),), torch.tensor([1., 2., 3.]))
+        tensor([0., 3., 0.])
+        >>> a.cuda().index_put_((torch.tensor([1, 1, 1]).cuda(),), torch.tensor([1., 2., 3.]).cuda())
+        tensor([0., 1., 0.], device='cuda:0')
+
     Args:
```
Comment (mruberry) [inline torch/__init__.py:459]: Typically args comes before examples

**Review by mruberry — COMMENTED** [review]
Hey @kurtamohler, overall this looks great. I made a few comments with some questions/suggestions but nothing major. Looking forward to hearing your thoughts!

File: torch/__init__.py (hunk 2 above)
Comment (kurtamohler) [reply torch/__init__.py:452]: Actually I just realized I was wrong about `embedding_bag`. It's not nondeterministic, it just has strange behavior if it's given an `offsets` input that doesn't make sense. I'll file an issue for that regardless.

File: torch/__init__.py (hunk 5 above)
Comment (kurtamohler) [reply torch/__init__.py:431]: Right, I'll only mention overlaps in this note.

Regarding order of operations, if the implementation of a given function could give a different order each time it's called in a consistent environment, then we should add an alert for it or offer an alternative implementation. We're already doing this--we considering `gpuAtomicAdd` nondeterministic for this exact reason.

For functions that do have a consistent internal order of operations, the order of operations is controlled by the user's code. I don't think we should alert in this case (I'm not even sure if/how we could detect something like this). In the following example, we're specifically asking for the sum to be performed with two different orders, and we get two different results. In short, if an order of operations difference is due to an actual user code change, it should still be considered deterministic.
```
>>> a = torch.randn(10000)
>>> a.sum().item()
6.353595733642578
>>> torch.stack([a[5000:], a[:5000]]).sum().item()
6.353599548339844
```

There's also the case of vectorized vs. nonvectorized operations, which effectively have different internal orders of operations. However, this is an environmental difference, so it should be considered deterministic as well.

Perhaps the reproducibility doc could mention some of these points (it already specifically mentions environmental differences, but not code differences), but I don't think the `use_deterministic_algorithms` doc is the right place for it. That is, unless there are some cases I'm not thinking of.

File: torch/__init__.py (hunk 6 above)
Comment (kurtamohler) [reply torch/__init__.py:446]: We should definitely not use `index_put_` since I'm going to add a nondeterministic alert for it.

The trouble is that many operations are now checking for overlaps and throwing an error, so it might be hard to find an example of a simple operation:
```
>>> a = torch.tensor([1.]).as_strided((3,), (0,))
>>> a.add_(torch.tensor([1., 2., 3.]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.
```

More operations will likely check this as time goes on. Actually, late last year I was working on a PR to add it to every inplace and out operation in codegen, but I took a break on it and haven't picked it back up yet. Since the goal is to have every out and inplace operation check for overlaps, perhaps we should consider not including an example at all since any currently working example is bound to go out of date at some point?

File: torch/__init__.py (hunk 8 above)
Comment (kurtamohler) [reply torch/__init__.py:443]: By "writing to multiple locations", I assume you're talking about duplicate indices. Since they will throw nondeterminism errors, I don't think we need an example anymore.

As mentioned above (https://github.com/pytorch/pytorch/pull/55413#discussion_r609071768), it will be difficult to maintain a working example for overlapped tensors as time goes on. Perhaps we shouldn't include any examples then? Or at least we could think about it an add them in another PR.

Comment (mruberry) [reply torch/__init__.py:443]: Yes, sorry, "writing to multiple locations" was the opposite of what I meant. I edited my comment for clarity.

Removing the overlapped example sounds good.

File: torch/__init__.py (hunk 6 above)
Comment (mruberry) [reply torch/__init__.py:446]: Your analysis makes a lot of sense. Sounds good.

File: torch/__init__.py (hunk 5 above)
Comment (mruberry) [reply torch/__init__.py:431]: I appreciate your thinking. Not elaborating here sounds good. Maybe we can think how to incorporate some ideas in the reproducibility note in the future.

**Comment by kurtamohler** [comment]
@mruberry, thanks for the comments! I think it's ready for another look

File: torch/__init__.py [hunk 11]
```diff
@@ -345,20 +345,20 @@ def set_default_dtype(d):
     """
     _C._set_default_dtype(d)
 
-def use_deterministic_algorithms(d):
+def use_deterministic_algorithms(mode):
     r""" Sets whether PyTorch operations must use "deterministic"
     algorithms. That is, algorithms which, given the same input, and when
     run on the same software and hardware, always produce the same output.
-    When True, operations will use deterministic algorithms when available,
+    When enabled, operations will use deterministic algorithms when available,
```
Comment (kurtamohler) [inline torch/__init__.py:352]: I figured "enabled" makes more sense since we're not explicitly referencing `mode` here

File: torch/__init__.py [hunk 12]
```diff
@@ -424,13 +424,36 @@ def use_deterministic_algorithms(d):
         * :func:`torch.bmm`
 
     Note that deterministic operations tend to have worse performance than
-    non-deterministic operations.
+    nondeterministic operations.
```
Comment (kurtamohler) [inline torch/__init__.py:427]: I noticed that sometimes we had "non-deterministic" and sometimes "nondeterministic". Better to be consistent

**Review by mruberry — APPROVED** [review]
Awesome! This looks great; thanks @kurtamohler!

**Comment by mruberry** [comment]
Hey @kurtamohler, sorry to bother you but for some reason the internal build system can't patch this onto the latest commit. Would you rebase this?
